internal/generate/olm-catalog/descriptor/descriptor.go

```diff
@@ -119,16 +150,57 @@ func GetCRDDescriptionForGVK(apisDir string, gvk schema.GroupVersionKind) (olmap
 	return crdDesc, nil
 }
 
-// getTypesFromDir gets all Go types from dir.
-func getTypesFromDir(dir string) (types.Universe, error) {
+func isDirExist(path string) (bool, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return fileInfo.IsDir(), nil
+}
+
+// getExpectedPkgLayout checks the directory layout in apisDir
+// for single and multi group layouts and returns the expected pkg path
+// for the group and version.
+// Returns empty string if neither single or multi group layout is detected
+// multi group path: <api-dir>/<group>/<version>
+// single group path: <api-dir>/<version>
+func getExpectedPkgLayout(apisDir, group, version string) (expectedPkgPath string, err error) {
+	groupVersionDir := filepath.Join(apisDir, group, version)
+	if isMultiGroupLayout, err := isDirExist(groupVersionDir); isMultiGroupLayout {
+		if err != nil {
+			return "", err
+		}
+		return groupVersionDir, nil
+	}
+	versionDir := filepath.Join(apisDir, version)
+	if isSingleGroupLayout, err := isDirExist(versionDir); isSingleGroupLayout {
+		if err != nil {
+			return "", err
+		}
+		return versionDir, nil
+	}
+	// Neither multi nor single group layout
+	return "", nil
+}
+
+// getTypesFromDir gets all Go types from dir and recursively its sub directories.
+// dir must be the project relative path to the pkg directory
+func getTypesFromDirRecursive(dir string) (types.Universe, error) {
 	if _, err := os.Stat(dir); err != nil {
 		return nil, err
 	}
-	if !filepath.IsAbs(dir) && !strings.HasPrefix(dir, ".") {
-		dir = fmt.Sprintf(".%s%s", string(filepath.Separator), dir)
-	}
+
+	// Gengo's AddDirRecursive fails to load subdir pkgs if the root dir
+	// isn't the full pkg import path.
+	dirImportPath := path.Join(projutil.GetGoPkg(), filepath.ToSlash(dir))
```

Comment:
What is wrong with using `dir` directly?

Reply:
Can't use `dir` directly since gengo complains it doesn't begin with `./`.
So did a bit of testing and seems like I can use `./<dir>` for AddDirRecursive() and it will work fine.
I've made the changes [here](https://github.com/operator-framework/operator-sdk/pull/2511/commits/655314e6864cc7554f833bd22b70eaa2dc350956#diff-56681196d4230f6aa4716b9c2ee4851e) and that's simplified a lot filepaths in the unit tests as well. Plus now we don't have to rely on constructing the full pkg import paths.